### ISSUE TITLE
CA1806 should not warn on TryParse where out parameter is used

### ISSUE THREAD

**Issue by Sti2nd** [dotnet/roslyn-analyzers#5919]
### Analyzer

**Diagnostic ID**: [CA1806](https://docs.microsoft.com/en-gb/dotnet/fundamentals/code-analysis/quality-rules/ca1806)

### Describe the improvement

For the [TryParse](https://docs.microsoft.com/en-us/dotnet/api/system.boolean.tryparse?view=netcore-3.1) methods that have an out parameter CA1806 should not warn if the out parameter is used.

### Describe suggestions on how to achieve the rule

Static code analysis? I assume it is possible to find out if a parameter is used or not.

### Additional context

In the following example CA1806 warns on the line with TryParse:

```csharp
var onboardedQueryParam = req.Query["onboarded"];
bool.TryParse(onboardedQueryParam, out var onlyOnboarded);

if (onlyOnboarded) {
    // onboardedQueryParam was parsed successfully and its value was equal to TrueString
} else {
    // onlyOnboarded is false here because onboardedQueryParam was equal to FalseString or because it is the default value.
}
```

The warning one get is:

> Run calls TryParse but does not explicitly check whether the conversion succeeded. Either use the return value in a conditional statement or <b>verify that the call site expects that the out argument will be set to the default value when the conversion fails.</b>

It seems from the message that the reason a warning is shown in the example is to instruct the developer to expect that the out parameter can have a default value. There is no need to show a warning to tell developers that variables have default values! 

One could argue that this rule is solely concerned about the return value of TryParse and not the out parameter, but in that case why is the message mentioning the out parameter? The best fix in my opinion is to detect the usage of the out parameter.

**Comment by FlukeFan** [comment]
Don't know if this helps, but the workaround I used was to use a discard parameter:

```
var onboardedQueryParam = req.Query["onboarded"];
var _ = bool.TryParse(onboardedQueryParam, out var onlyOnboarded);
```

**Comment by JinsPeter** [comment]
why not use a discard? Discard is introduced for intentionally ignoring method values.
but in your case, Try to be more explicit about implementation

```csharp
if(bool.TryParse(onboardedQueryParam, out var onlyOnboarded)  && onlyOnboarded )
{
   //onboardedQueryParam was parsed successfully and its value was equal to TrueString
} 
else {
    // Do when you have a false or has nothing


```
